### PR TITLE
add isl_aff_eval

### PR DESCRIPTION
**Pull request by skimo-openhub**
There is some code in TC that begging to be replaced by a call to isl_aff_eval.
Provide this function and clean up some stuff along the way.

### CODE REVIEW

File: isl_aff.c [hunk 1]
```diff
@@ -346,9 +346,16 @@ int isl_aff_find_dim_by_name(__isl_keep isl_aff *aff, enum isl_dim_type type,
 	return isl_local_space_find_dim_by_name(aff->ls, type, name);
 }
 
+/* Return the domain space of "aff".
+ */
+__isl_keep isl_space *isl_aff_peek_domain_space(__isl_keep isl_aff *aff)
```
Comment (ftynse) [inline isl_aff.c]: this function is not declared in a header, may be make it `static`?

File: isl_local.c [hunk 2]
```diff
@@ -182,3 +183,48 @@ int isl_local_cmp(__isl_keep isl_local *local1, __isl_keep isl_local *local2)
 
 	return 0;
 }
+
+/* Extend a vector "v" representing an integer point
```
Comment (ftynse) [inline isl_local.c]: does a vector that represents an integer point always start with a denominator `1`?  (there's a check in line 210)

File: isl_aff.c [hunk 3]
```diff
@@ -9394,3 +9395,91 @@ __isl_give isl_multi_pw_aff *isl_multi_union_pw_aff_extract_multi_pw_aff(
...
+	isl_ctx *ctx;
+	isl_val *v;
+
+	if (!aff || !pnt)
+		return NULL;
+
+	ctx = isl_vec_get_ctx(aff);
+	isl_int_init(n);
+	isl_int_init(d);
+	isl_seq_inner_product(aff->el + 1, pnt->el, pnt->size, &n);
```
Comment (ftynse) [inline isl_aff.c]: if `pnt` starts with a denominator as the comment says, shouldn't the inner product start at `pnt->el + 1` instead?

**Comment by skimo-openhub** [comment]
On Thu, Feb 22, 2018 at 02:23:19PM +0000, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -346,9 +346,16 @@ int isl_aff_find_dim_by_name(__isl_keep isl_aff *aff, enum isl_dim_type type,
>  	return isl_local_space_find_dim_by_name(aff->ls, type, name);
>  }
>  
> +/* Return the domain space of "aff".
> + */
> +__isl_keep isl_space *isl_aff_peek_domain_space(__isl_keep isl_aff *aff)
> 
> this function is not declared in a header, may be make it `static`?

I expect this function may become useful in other TUs,
but, you are right, I should make it static until
it actually gets exposed.

skimo

**Comment by skimo-openhub** [comment]
On Thu, Feb 22, 2018 at 02:33:40PM +0000, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -182,3 +183,48 @@ int isl_local_cmp(__isl_keep isl_local *local1, __isl_keep isl_local *local2)
>  
>  	return 0;
>  }
> +
> +/* Extend a vector "v" representing an integer point
> 
> does a vector that represents an integer point always start with a denominator `1`?  (there's a check in line 210)

Yes.

skimo

**Comment by skimo-openhub** [comment]
On Thu, Feb 22, 2018 at 02:40:15PM +0000, ftynse wrote:
> ftynse commented on this pull request.
> > + * in the coordinates (with denominator) "pnt".
> + */
> +static __isl_give isl_val *eval(__isl_keep isl_vec *aff,
> +	__isl_keep isl_vec *pnt)
> +{
> +	isl_int n, d;
> +	isl_ctx *ctx;
> +	isl_val *v;
> +
> +	if (!aff || !pnt)
> +		return NULL;
> +
> +	ctx = isl_vec_get_ctx(aff);
> +	isl_int_init(n);
> +	isl_int_init(d);
> +	isl_seq_inner_product(aff->el + 1, pnt->el, pnt->size, &n);
> 
> if `pnt` starts with a denominator as the comment says, shouldn't the inner product start at `pnt->el + 1` instead?

No, the constant term of the affine expression should
be multiplied by the denominator because the other terms
are multiplied by the numerators.
Note that this piece of code does handle rational points.
In any case, the constant term cannot simply be left out.

skimo

**Pull request merged**